Add documentation for operation-mode in KPOps (#565)
bakdata/kpops: docs/docs/user/core-concepts/operation-mode.md

# Operation Modes in KPOps

KPOps supports three operation modes—`managed`, `manifest`, and `argo`. These modes determine how resources are managed and allow users to tailor their deployment strategy.

- **Managed Mode** (default): KPOps uses Helm, and communicates with services like Kafka Rest Proxy, and Kafka Connect under the hood to manage the installation/(graceful) deletion of applications, creation/deletion of Kafka topics, creation/deletion of Connectors defined in your `pipeline.yaml`.
- **Manifest Mode**: Focuses on generating Kubernetes manifests.
- **Argo Mode**: Extends the functionality to include ArgoCD-specific hooks for certain operations, facilitating GitOps workflows with automated cleanup and reset tasks.

---

## Configuring Operation Modes

> It is possible to configure the operation mode in the `config.yaml` file. Please refer to the [Configuration documentation page](https://bakdata.github.io/kpops/9.0/user/core-concepts/config/).

Reviewer: will links automatically update the version when we do another major release in the future? otherwise I would remove the version number from the link

Author: It is updated!

Reviewer: this one is not though

Alternatively, you can to pass the `--operation-mode <OPERATION>` option in the CLI to override the operation mode of the `config.yaml`. You can refer to the [CLI commands documentation](https://bakdata.github.io/kpops/9.0/user/references/cli-commands/#kpops-deploy) for more details.

---

## Generated Resources by Mode and Operation

### `deploy`

#### **Manifest Mode**

- **streams-bootstrap Applications**:
  - Depending on your pipeline configuration, Kubernetes `Job`, `Deployment`, `ConfigMap`, and `Service` resources.
  - Please refer to [Streams-bootstrap Helm Charts](https://github.com/bakdata/streams-bootstrap/tree/master/charts).
- **Topics**:
  - Strimzi `KafkaTopic` CRDs.

#### **Argo Mode**

- **streams-bootstrap Applications**:
  - Depending on your pipeline configuration, Kubernetes `Job`, `Deployment`, `ConfigMap`, and `Service` resources.

>   - Additional `sync-wave` annotation with value of `>0` to prioritizes Kafka Topics deployment over the application

Reviewer: this reads like higher number means higher priority but it's the other way around, right?

we could avoid possible confusion by saying: 

```suggestion
  - Additional Argo `sync-wave` annotation to ensure Kafka topics are created first (default `sync-wave=0`) before deploying apps (lower priority `sync-wave>0`). All components of each sync wave are deployed in parallel by Argo.
```

>   - Please refer to [Streams-bootstrap Helm Charts](https://github.com/bakdata/streams-bootstrap/tree/master/charts).

Reviewer: found one outlier 😛 
```suggestion
  - Please refer to [streams-bootstrap Helm Charts](https://github.com/bakdata/streams-bootstrap/tree/master/charts).
```

- **Topics**:
  - Strimzi `KafkaTopic` CRDs.
- **Cleanup Jobs**:
  - Kubernetes `Job` resources configured **with** ArgoCD `PostDelete` hooks, ensuring cleanup tasks are executed after ArgoCD application deletion.

---

### `reset`

#### **Manifest Mode**

- **Topics**:
  - Strimzi `KafkaTopic` CRDs.
- **Reset Jobs**:
  - Kubernetes `Job` resources for resetting Kafka Streams application states.

#### **Argo Mode**

- **Topics**:
  - Strimzi `KafkaTopic` CRDs.
- **Reset Jobs**:
  - Kubernetes `Job` resources **without** ArgoCD `PostDelete` hooks, providing a simpler reset process.

---

### `clean`

#### **Manifest Mode**

- **Clean Jobs**:
  - Kubernetes `Job` resources for cleaning up temporary resources or artifacts using application container images.

#### **Argo Mode**

- **Not Applicable**:

>   - The `clean` command is not supported in Argo mode. The clean can be achieved through manifesting hooked cleanup jobs during the `deploy` command.

Reviewer: ```suggestion
  - The `clean` command is not supported in Argo mode. The clean is instead achieved through cleanup job hooks during the `deploy` command.
```

# Operation Modes in KPOps

KPOps supports three operation modes—`managed`, `manifest`, and `argo`. These modes determine how resources are managed and allow users to tailor their deployment strategy.

> - **Managed Mode** (default): KPOps uses helm under the hood to manage the installation of applications in the defined pipeline.yaml.

Reviewer: ```suggestion
- **Managed Mode** (default): KPOps uses Helm under the hood to manage the installation of applications in the defined pipeline.yaml.
```

Reviewer: not just Helm, we talk to all the necessary services and deploy directly to the cluster. That's the main difference to Manifest mode, which also uses Helm under the hood, but only for `helm template`.

- **Manifest Mode**: Focuses on generating Kubernetes manifests.
- **Argo Mode**: Extends the functionality to include ArgoCD-specific hooks for certain operations, facilitating GitOps workflows with automated cleanup and reset tasks.

---

## Configuring Operation Modes

### `manifest` Mode

Enable `manifest` mode by setting the following in your configuration:

```yaml
operation_mode: manifest
```

### `argo` Mode

Enable `argo` mode by setting the following in your configuration:

```yaml
operation_mode: argo

> ```

Reviewer: this is a bit verbose imo. It would be enough to link to the configuration page

Author: Re-written and linked the configuration page

<!-- dprint-ignore-start -->

!!! info Set operation-mode over CLI
     It is possible to pass the `--operation-mode <OPERATION>` option in the CLI to override the operation mode of the `config.yaml`. You can refer to the [CLI commands documentation](https://bakdata.github.io/kpops/9.0/user/references/cli-commands/#kpops-deploy) for more details.

<!-- dprint-ignore-end -->

---

## Generated Resources by Mode and Operation

### `deploy`

#### **Manifest Mode**

- **Streams bootstrap Applications**:
  - Depending on your pipeline configuration, Kubernetes `Job`, `Deployment`, `ConfigMap`, and `Service` resources.
  - Please refer to [Streams-bootstrap Helm Charts](https://github.com/bakdata/streams-bootstrap/tree/master/charts).
- **Topics**:
  - Strimzi `KafkaTopic` CRDs.

#### **Argo Mode**

> - **Streams bootstrap Applications**:

Reviewer: please stick to one consistent naming, I believe the official name is _streams-bootstrap_
